### PR TITLE
Json server

### PR DESCRIPTION
**Pull request by maciejkocur**

### CODE REVIEW

File: server/src/main/java/io/github/expansionteam/battleships/JsonHandler.java [hunk 1]
```diff
@@ -58,23 +63,24 @@ private String getOpponentArrivedEventJson(JSONObject jsonResponse) {
 
     private String getShipsGeneratedEventJson(JSONObject jsonResponse, Collection<Ship> ships) {
         JSONArray shipsJson = new JSONArray();
+        JSONObject position;
+        JSONObject shipJson;
```
Comment (clinton56) [inline server/src/main/java/io/github/expansionteam/battleships/JsonHandler.java]: `shipJson` is almost like `shipsJson`. Are you aware that now it is very hard to recognize between these 2 objects?

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java [hunk 2]
```diff
@@ -19,6 +19,7 @@ private Board(BoardBuilder bb) {
     public static class BoardBuilder {
 
         private final Set<Field> board = initializeSet();
+
```
Comment (clinton56) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: Unnecessary empty line :ghost:

File: server/src/main/java/io/github/expansionteam/battleships/JsonHandler.java (hunk 1 above)
Comment (maciejkocur) [inline server/src/main/java/io/github/expansionteam/battleships/JsonHandler.java]: changed

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java (hunk 2 above)
Comment (maciejkocur) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: fixed

File: server/src/main/java/io/github/expansionteam/battleships/Game.java [hunk 3]
```diff
@@ -52,4 +53,18 @@ public boolean shoot(int x, int y) {
         secondPlayerBoard.shootField(x, y);
         return isFieldHit(secondPlayerBoard, x, y);
     }
+
+    public boolean isDestroyedShip(int x, int y) {
+        if (firstPlayer()) {
```
Comment (LIttleAncientForestKami) [inline server/src/main/java/io/github/expansionteam/battleships/Game.java]: What do you say to: `firstPlayer() : return firstPlayerBoard ? return secondPlayer`

File: server/src/main/java/io/github/expansionteam/battleships/Game.java [hunk 4]
```diff
@@ -52,4 +53,18 @@ public boolean shoot(int x, int y) {
...
+            return firstPlayerBoard.isDestroyedShip(x, y);
+        }
+        return secondPlayerBoard.isDestroyedShip(x, y);
+    }
+
+    public Collection<Field> getAdjacentToShip(int x, int y) {
+        if (firstPlayer()) {
```
Comment (LIttleAncientForestKami) [inline server/src/main/java/io/github/expansionteam/battleships/Game.java]: As above

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java [hunk 5]
```diff
@@ -86,6 +85,24 @@ public boolean shootField(int x, int y) {
...
+    public Collection<Field> getAdjacentToShip(int x, int y) {
+        Field field = getFieldFromTheBoard(x, y);
+        Ship parentShip = field.getParentShip();
+        if (parentShip == null) {
+            return Collections.emptySet();
+        }
+        return Ship.generateSetOfAdjacentFields(this, parentShip.occupiedFields);
+    }
+
     // when the ship is destroyed
```
Comment (LIttleAncientForestKami) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: Method comment -> Javadoc.

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java [hunk 6]
```diff
@@ -86,6 +85,24 @@ public boolean shootField(int x, int y) {
...
+        return parentShip.isDestroyed();
+    }
+
+    public Collection<Field> getAdjacentToShip(int x, int y) {
+        Field field = getFieldFromTheBoard(x, y);
+        Ship parentShip = field.getParentShip();
+        if (parentShip == null) {
+            return Collections.emptySet();
+        }
+        return Ship.generateSetOfAdjacentFields(this, parentShip.occupiedFields);
```
Comment (LIttleAncientForestKami) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: Nice thinking.

File: server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java [hunk 7]
```diff
@@ -131,17 +131,20 @@ public void responsesToHitShip() {
 
         Game gameMock = mock(Game.class);
         when(gameMock.shoot(3, 5)).thenReturn(true);
+
+        JSONObject position = new JSONObject()
+                .put("position", new JSONObject().put("x", 3).put("y", 5));
         String requestJson = new JSONObject()
                 .put("type", "ShootPositionEvent")
-                .put("position", new JSONObject().put("x", 3).put("y", 5))
+                .put("data", position)
```
Comment (LIttleAncientForestKami) [inline server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java]: Good change.

File: server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java [hunk 8]
```diff
@@ -150,29 +153,101 @@ public void responsesToHitShip() {
     }
 
     @Test
-    public void responsesToHitEmtyField() {
+    public void responsesToHitEmptyField() {
```
Comment (LIttleAncientForestKami) [inline server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java]: :+1:

File: server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java [hunk 9]
```diff
@@ -150,29 +153,101 @@ public void responsesToHitShip() {
...
+        expectedAdjacentCoordinates.add(3);
+        expectedAdjacentCoordinates.add(5);
+        expectedAdjacentCoordinates.add(1);
+        expectedAdjacentCoordinates.add(6);
+        expectedAdjacentCoordinates.add(2);
+        expectedAdjacentCoordinates.add(6);
+        expectedAdjacentCoordinates.add(3);
+        expectedAdjacentCoordinates.add(6);
+
+        // when
```
Comment (LIttleAncientForestKami) [inline server/src/test/java/io/github/expansionteam/battleships/JsonHandlerTest.java]: Your when contains `given` code.
You want builders in these tests: position builder, actual response and shootPosition. Code for them gets repeated and litters your test logic.

File: server/src/test/java/io/github/expansionteam/battleships/engine/BoardTest.java [hunk 10]
```diff
@@ -190,4 +187,97 @@ public void marksAdjacentFieldsIfShipsDestroyed(Field field, Orientation orienta
             assertEquals(adjacentField.isHit(), true);
         }
     }
+
+    @Test
+    public void isDestroyed() {
```
Comment (LIttleAncientForestKami) [inline server/src/test/java/io/github/expansionteam/battleships/engine/BoardTest.java]: Board test is destroyed seems to NOT be about ship. `willDestroyGivenShip` perhaps?

Comment (LIttleAncientForestKami) [inline server/src/test/java/io/github/expansionteam/battleships/engine/BoardTest.java]: same with other methods in this test class.

File: server/src/main/java/io/github/expansionteam/battleships/Game.java (hunk 3 above)
Comment (KPP13) [inline server/src/main/java/io/github/expansionteam/battleships/Game.java]: We have already done it on later commits :)

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java (hunk 5 above)
Comment (KPP13) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: Javadoc will be added soon.

File: server/src/main/java/io/github/expansionteam/battleships/engine/Board.java (hunk 6 above)
Comment (KPP13) [inline server/src/main/java/io/github/expansionteam/battleships/engine/Board.java]: But also have to be changed to ? :

**Pull request merged**